File: programming_methods/laboratories/6/5.cpp
/**
 * task b aka 2781
 * link: http://informatics.mccme.ru/moodle/mod/statements/view3.php?id=1974&chapterid=2781#1
 * Copyright reserved.
 * author: Gumeniuk Stanislav
 */

#include <iostream>
#include <fstream>
#include <time.h>
#include <stdlib.h>

using namespace std;

ifstream in("input.txt");
ofstream out("output.txt");

class vertex{
    public:
    int x, y, label;
    struct vertex *parent, *left, *right;

    vertex() {}

    void set(int xi, int yi, int labeli){
        x = xi;
        y = -yi;
        label = labeli;
        parent = NULL;
        left = NULL;
        right = NULL;
    }

    vertex& operator=(const vertex& a){
        x = a.x;
        y = a.y;
        label = a.label;
        parent = parent;
        left = left;
        right = right;
        return *this;
    }

    friend void swap (vertex& a, vertex& b);
};

void swap (vertex& a, vertex& b){
    vertex shift;
    shift = a;
    a = b;
    b = shift;
}

typedef vertex *pvertex;


class decTree{
    private:
    int treeSize;
    vertex *vertexs;
    pvertex *ordered;
    pvertex root;
    int vertexCount;

    void _sortIns(pvertex *pvertexs , int left, int right){
        pvertex shift;

        for (int i = left + 1; i < right; i++){
            int j = i - 1;
            shift = pvertexs[i];

            for (int k = i-1; ((k >= left) && (pvertexs[j]->x > shift->x)) ; k--, j--){
                pvertexs[k + 1] = pvertexs[k];
            }
            pvertexs[j + 1] = shift;
        }
    }

    void _sortQuick(pvertex *pvertexs, int left, int right){
        srand(time(NULL));
        int l = left;
        int r = right-1;
        pvertex p;

        p = pvertexs[l + rand() % (right - left)];

        do{
            while (pvertexs[l]->x < p->x) {
                l++;
            }
            while (pvertexs[r]->x > p->x) {
                r--;
            }

            if (l <= r){
                swap(pvertexs[l++], pvertexs[r--]);
            }
        } while (l <= r);

        if (r - left > 16){
            _sortQuick(pvertexs, left, r + 1);
        } else if (r > left){
            _sortIns(pvertexs, left, r + 1);
        }

        if (right - l > 16){
            _sortQuick(pvertexs, l, right);
        } else if (l < right - 1){
            _sortIns(pvertexs, l , right);
        }
        ////

    }

    void _makeTree(){
        pvertex last = NULL;
        root = ordered[0];
        last = root;
        for (int i = 1; i < treeSize; i++){
            while (last->parent != NULL && ordered[i]->y > last->y){
                last = last->parent;
            }
            if (ordered[i]->y <= last->y){
                ordered[i]->left = last->right;
                ordered[i]->parent = last;
                if (last->right){
                    last->right->parent = ordered[i];
                }
                last->right = ordered[i];

            } else {
                ordered[i]->left = last;
                last->parent = ordered[i];
                root = ordered[i];
            }
            last = ordered[i];
        }

    }



    public:

    void vertexAdd(int x, int y){
        vertexs[vertexCount].set(x, y, vertexCount + 1);
        ordered[vertexCount] = vertexs + vertexCount;
        vertexCount++;
        _sortQuick(ordered, 0, treeSize);
        _makeTree();
    }

    decTree(int size):treeSize(size){
        vertexCount = 0;
        vertexs = new vertex[treeSize];
        ordered = new pvertex[treeSize];
    }

    decTree(){
        int a , b;
        vertexCount = 0;

        in >> treeSize;
        vertexs = new vertex[treeSize];
        ordered = new pvertex[treeSize];
        for (int i = 0; i < treeSize; i++ ){
            in >> a >> b;
            vertexs[vertexCount].set(a, b, vertexCount + 1);
            ordered[vertexCount] = vertexs + vertexCount;
            vertexCount++;
        }
        _sortQuick(ordered, 0, treeSize);
        _makeTree();
    }




    void resultsDisplay(){
        int parent, left, right;
        out << "YES" << endl;
        for (int i = 0; i < treeSize; i++){
            if (vertexs[i].parent) {
                parent = vertexs[i].parent->label;
            } else {
                parent = 0;
            }

            if (vertexs[i].left){
                left = vertexs[i].left->label;
            }  else {
                left = 0;
            }

            if (vertexs[i].right) {
                right = vertexs[i].right->label;
            } else {
                right = 0;
            }

            out << parent << " " << left << " " << right << endl;
        }
    }
};




int main(void){
    decTree *mytree = new decTree();
    mytree->resultsDisplay();
    return 0;
}

File: programming_methods/laboratories/6/5.py
"""
Задача №2781. Декартово дерево

Вам даны пары чисел (ai,bi), Вам необходимо построить декартово дерево, такое что i-ая вершина имеет ключи (ai,bi),
вершины с ключом ai образуют бинарное дерево поиска, а вершины с ключом bi образуют кучу на минимум.

Входные данные

В первой строке записано число N — количество пар.
Далее следует N (1≤N≤50000) пар (ai,bi). Для всех пар |ai|,|bi|≤30000. ai≠aj и bi≠bj для всех i≠j.

Выходные данные

Если декартово дерево с таким набором ключей построить возможно, выведите в первой строке YES, в противном случае выведите NO.
В случае ответа YES, выведите N строк, каждая из которых должна описывать вершину.
Описание вершины состоит из трёх чисел: номер предка, номер левого сына и номер правого сына.
Если у вершины отсутствует предок или какой-либо из сыновей, то выводите на его месте число 0.

Если подходящих деревьев несколько, выведите любое.
"""

from dataclasses import dataclass
from typing import List, Optional, Sequence, Tuple, cast


@dataclass
class Vertex:
    """
    Класс для представления вершины декартова дерева.

    Атрибуты:
        x (int): Ключ для бинарного дерева поиска.
        y (int): Приоритет для свойства кучи.
        label (int): Уникальная метка вершины.
        parent (Optional[Vertex]): Родительская вершина.
        left (Optional[Vertex]): Левая дочерняя вершина.
        right (Optional[Vertex]): Правая дочерняя вершина.
    """

    x: int
    y: int
    label: int
    parent: Optional["Vertex"] = None
    left: Optional["Vertex"] = None
    right: Optional["Vertex"] = None


class DecTree:
    def __init__(self, pairs: List[Tuple[int, int]]) -> None:
        """
        Класс для представления декартова дерева.
        :param pairs: Список пар (x, y), где x — ключ, а y — приоритет.
        """
        self.__vertices = [
            Vertex(x=a, y=b, label=i + 1) for i, (a, b) in enumerate(pairs)
        ]
        self.ordered = sorted(self.vertices, key=lambda v: v.x)
        self.root: Optional[Vertex] = None

        self._build_tree()

    @property
    def vertices(self) -> Sequence[Vertex]:
        """
        Возвращает список всех вершин дерева.
        :returns: Возвращает список всех вершин дерева.
        """
        return self.__vertices

    def is_empty(self) -> bool:
        """
        Проверяет, пусто ли дерево.
        :returns: True, если дерево пусто, иначе False.
        """
        return len(self.vertices) == 0

    def _build_tree(self) -> None:
        """
        Строит декартово дерево, удовлетворяющее свойствам бинарного дерева поиска и кучи на минимум.

        Алгоритм:
            1. Упорядочиваем вершины по ключу x (для бинарного дерева поиска).
            2. Начинаем с первой вершины как корня.
            3. Для каждой следующей вершины ищем место в дереве:
                - Если приоритет вершины y больше, поднимаем её вверх, переставляя связи.
                - Если приоритет y меньше, добавляем вершину в правого потомка текущей.
            4. Поддерживаем свойства бинарного дерева поиска (по ключу x) и кучи (по y).
        """
        self.root = self.ordered[0]
        last = self.root

        for vertex in self.ordered[1:]:
            # Найти корректное место для новой вершины в декартовом дереве
            while last and last.parent and vertex.y < last.y:
                last = last.parent

            if vertex.y >= last.y:
                vertex.left = last.right
                vertex.parent = last
                if last.right:
                    last.right.parent = vertex
                last.right = vertex
            else:
                vertex.left = last
                last.parent = vertex
                self.root = vertex

            last = vertex

    def __str__(self) -> str:
        """
        Возвращает строковое представление дерева для вывода.

        Формат:
            Для каждой вершины выводятся три числа:
                1. Метка родителя (0, если родителя нет).
                2. Метка левого сына (0, если сына нет).
                3. Метка правого сына (0, если сына нет).

        :returns: Строка с описанием всех вершин дерева.
        """
        result: List[Tuple[int, int, int]] = [
            (
                v.parent.label if v.parent else 0,
                v.left.label if v.left else 0,
                v.right.label if v.right else 0,
            )
            for v in self.vertices
        ]

        return "\n".join(" ".join(map(str, collection)) for collection in result)


def main() -> None:
    n: int = int(input())
    pairs: List[Tuple[int, int]] = cast(
        List[Tuple[int, int]], [tuple(map(int, input().split())) for _ in range(n)]
    )
    tree: DecTree = DecTree(pairs)

    if tree.is_empty():
        print("NO")
    else:
        print("YES", tree, sep="\n")


if __name__ == "__main__":
    main()
